misc/vtk/src/processors/vtktovolume.cpp

```diff
@@ -0,0 +1,422 @@
+                if (dynamic_cast<BoolProperty *>(prop)->get())
+                    formerChannels.insert(prop->getDisplayName());
+        }
+    }
+
+    dataArrays_.clear();
+    if (!dataFormats_.size() || !inport_.hasData()) {
+        reactToChange_ = true;
+        return;
+    }
+    const VTKDataSet &vtkData = *inport_.getData();
+
+    auto selectedFormat = dataFormats_.getSelectedValue();
+    auto cellData = vtkData->GetCellData();
+    auto pointData = vtkData->GetPointData();
+    if ((useCellData_.get() && !cellData) || (!useCellData_.get() && !pointData)) return;
+    int numArrays =
+        useCellData_.get() ? cellData->GetNumberOfArrays() : pointData->GetNumberOfArrays();
+
+    int numTotalComps = 0;
+    for (int i = 0; i < numArrays; ++i) {
+        auto array = useCellData_.get() ? cellData->GetArray(i) : pointData->GetArray(i);
+        std::string arrayName(array->GetName());
+
+        int vtkTypeId = array->GetDataType();
+        const DataFormatBase *format =
+            vtkutil::getFormatFromVtkId(vtkTypeId, array->GetElementComponentSize() * 8);
+        if (format->getId() != selectedFormat) continue;
+        int numComps = array->GetNumberOfComponents();
+
+        numTotalComps += numComps;
+        dataArrays_.addArray(arrayName, numComps, array->GetVoidPointer(0));
+        if (keepSettings) {
+            BoolProperty *prop = dynamic_cast<BoolProperty *>(dataArrays_[dataArrays_.size() - 1]);
+            if (prop && formerChannels.find(prop->getDisplayName()) != formerChannels.end())
+                prop->set(true);
+        }
+    }
+    if (!keepSettings && numTotalComps <= 4)
+        for (auto *prop : dataArrays_.getProperties())
+            if (BoolProperty *boolProp = dynamic_cast<BoolProperty *>(prop)) boolProp->set(true);
+
+    dataArrays_.selectedFormat = dataFormats_.get();
+
+    if (keepSettings) convertData_ = true;
+
+    reactToChange_ = true;
+}
+
+void VTKtoVolume::updateAvailableArrays() {
+    NetworkLock lock;
+    reactToChange_ = false;
+
+    int numTotalComps = 0;
+    auto properties = dataArrays_.getProperties();
+    for (size_t p = 0; p < properties.size(); ++p) {
+        if (dynamic_cast<BoolProperty *>(properties[p])->get())
+            numTotalComps += dataArrays_.numArrayComponents[p];
+    }
+
+    for (size_t p = 0; p < properties.size(); ++p) {
+        if (dynamic_cast<BoolProperty *>(properties[p])->get()) continue;
+        bool tooMuch = numTotalComps + dataArrays_.numArrayComponents[p] > 4;
+        properties[p]->setReadOnly(tooMuch);
+    }
+
+    reactToChange_ = true;
+}
+
+void VTKtoVolume::convertData() {
+    outport_.setData(nullptr);
+    if (!inport_.hasData()) return;
+
+    auto vtkDataPtr = inport_.getData();
+
+    // Get dimensions. If none are given, it's not a regular grid.
+    std::optional<size3_t> dimensionsOpt = vtkDataPtr->getDimensions();
+    if (!dimensionsOpt) {
+        LogWarn("Input vtk data set is not rectilinear.");
```

Comment:
This should be an error.

Reply:
agree, throw an exception instead.

Reply:
That feels excessive: it only mean the data is not regular enough for this processor. I.e., the `vtkDataSet` we receive is not a `vtkStructuredGrid`, `vtkRectilinearGrid` or a version of `vtkImageData`, and thus has no 3D indexing.
The processor is then not able to create a volume, but the vtk data might be completely valid.